### PR TITLE
fix: new message in empty messagelist and channel preview latest message

### LINKED ISSUE #2452: [🐛] `EmptyStateIndicator` from `Channel` not hidden when new message is sent
## Issue

When passing `EmptyStateIndicator` to the `Channel` component (and a user is viewing a channel with no messages), sending a new message in the channel will **not** automatically hide the `EmptyStateIndicator` and show the `MessageList`.

Here's how we are rendering the `Channel` component:

```tsx
<Channel
  channel={channel}
  disableKeyboardCompatibleView
  initialScrollToFirstUnreadMessage
  autoCompleteTriggerSettings={autoCompleteTriggerSettings}
  markdownRules={markdownRules}
  messageActions={messageActions}
  myMessageTheme={myMessageTheme}
  UrlPreview={UrlPreview}
  EmptyStateIndicator={() => <EmptyState channel={channel} />}
>
  <MessageList onListScroll={dismissKeyboard} />
  <MessageInput />
</Channel>
```

and here's how we are rendering out `EmptyState` with a predefined message example that a user can send in the channel:

```tsx
const EmptyState = ({ channel }: Readonly<{ channel: ChannelType }>) => {
  const { mutate: sendMessage } = useMutation({
    mutationFn: (text: string) => channel.sendMessage({ text }),
  });

  return (
    <VStack>
      <Text>{'No messages. Example messages you can send:'}</Text>
      <Button
        onPress={() => {
          sendMessage('What is the weather today?');
        }}
      >
        {'What is the weather today?'}
      </Button>
    </VStack>
  );
};
```

However I've tested with sending a message from a different user account and this issue is also present there - I can see the typing indicator and then when the message is sent the `EmptyStateIndicator` is still shown.

### Steps to reproduce

1. Setup a `Channel` component with an `EmptyStateIndicator`
2. View a channel with no messages
3. Send a new message in the channel with `channel.sendMessage` or by sending a new message from another user
4. Observe that the `EmptyStateIndicator` is not hidden

### Expected behavior

The `EmptyStateIndicator` is automatically hidden and the `MessageList` is shown.

### Customization

<details><summary>Click To Expand</summary>
<p>

```typescript jsx
# N/A
```
</p>
</details>

### Offline support

- [ ] I have enabled offline support.
- [ ] The feature I'm having does not occur when offline support is disabled. (stripe out if not applicable)

## Environment

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
"stream-chat-expo": "^5.25.0",
```

**`react-native info` output:**

```
System:
  OS: macOS 14.3.1
  CPU: (10) arm64 Apple M1 Max
  Memory: 58.64 MB / 32.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 18.18.2
    path: ~/.nvm/versions/node/v18.18.2/bin/node
  Yarn:
    version: 4.1.1
    path: ~/.nvm/versions/node/v18.18.2/bin/yarn
  npm:
    version: 9.8.1
    path: ~/.nvm/versions/node/v18.18.2/bin/npm
  Watchman:
    version: 2024.01.22.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /opt/homebrew/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.4
      - iOS 17.4
      - macOS 14.4
      - tvOS 17.4
      - visionOS 1.1
      - watchOS 10.4
  Android SDK:
    API Levels:
      - "33"
    System Images:
      - android-33 | Google Play ARM 64 v8a
    Android NDK: Not Found
IDEs:
  Android Studio: 2023.2 AI-232.10227.8.2321.11479570
  Xcode:
    version: 15.3/15E204a
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 21.0.2
    path: /usr/bin/javac
  Ruby:
    version: 2.6.10
    path: /usr/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react: Not Found
  react-native: Not Found
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: Not found
  newArchEnabled: Not found
iOS:
  hermesEnabled: Not found
  newArchEnabled: Not found
```

- **Platform that you're experiencing the issue on**:
    - [x] iOS
    - [ ] Android
    - [x] **iOS** but have not tested behavior on Android
    - [ ] **Android** but have not tested behavior on iOS
    - [ ] Both
- **`stream-chat-react-native` version you're using that has this issue:**
    - `5.25.0`
- Device/Emulator info:
  - [x] I am using a physical device
  - OS version: `iOS 174`
  - Device/Emulator: `iPhone 13 Pro`

</p>
</details>

### Screenshots

<details><summary>Click To Expand</summary>
<p>

Here's an example video:

https://github.com/GetStream/stream-chat-react-native/assets/712727/4d6c91b8-e965-4a86-ab15-1ef55cfd9a82

and logs to indicate that a new message is sent when pressing the button:

```
 LOG  Sentry Logger [log]: [TouchEvents] Touch event within element: View
 LOG  Sentry Logger [log]: [TouchEvents] Touch event within element: Text
 INFO  stream-chat: client: post - Request - https://chat.stream-io-api.com/channels/messaging/7d2ebf07-dfcf-4efb-8171-d288c29fb866/message
 INFO  stream-chat: connection:onmessage() - onmessage callback
 INFO  stream-chat: client:_handleClientEvent - Received event of type { message.new }
 INFO  stream-chat: channel:_handleChannelEvent - Received event of type { message.new } on messaging:7d2ebf07-dfcf-4efb-8171-d288c29fb866
 INFO  stream-chat: client:post - Response - url: https://chat.stream-io-api.com/channels/messaging/7d2ebf07-dfcf-4efb-8171-d288c29fb866/message > status 201
 INFO  stream-chat: connection:onmessage() - onmessage callback
 INFO  stream-chat: client:_handleClientEvent - Received event of type { health.check }
```

when manually navigating back and then into the message screen the new message is then visible:

</p>
</details>

---

### PR DESCRIPTION
**Pull request by khushal87**
## 🎯 Goal

This PR fixes #2452, and the channel preview message in the channel list when the list is empty.

## 🎨 UI Changes

<details>
<summary>iOS</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

<details>
<summary>Android</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

## ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [ ] Documentation is updated
- [x] New code is tested in main example apps, including all possible scenarios
  - [x] SampleApp iOS and Android
  - [x] Expo iOS and Android

### CODE REVIEW

File: package/src/components/Channel/Channel.tsx [hunk 1]
```diff
@@ -736,6 +732,7 @@ const ChannelWithContext = <
   const copyMessagesState = useRef(
     throttle(
       () => {
+        console.log(channel.state.messageSets, channel.state.isUpToDate);
```
Comment (santhoshvai) [inline package/src/components/Channel/Channel.tsx:735]: remove this log pls

File: package/src/components/ChannelList/hooks/listeners/useNewMessage.ts [hunk 2]
```diff
@@ -34,27 +35,22 @@ export const useNewMessage = <
       } else {
         setChannels((channels) => {
           if (!channels) return channels;
+          const channelInList = channels.filter((channel) => channel.cid === event.cid).length > 0;
```
Comment (santhoshvai) [inline package/src/components/ChannelList/hooks/listeners/useNewMessage.ts:38]: why move this before the if though? 

if lock was enabled.. we wouldnt need to filter

Comment (santhoshvai) [reply package/src/components/ChannelList/hooks/listeners/useNewMessage.ts:38]: or was that behaviour a bug?

File: package/src/components/ChannelList/hooks/listeners/useNewMessage.ts [hunk 3]
```diff
@@ -34,27 +35,22 @@ export const useNewMessage = <
...
+          if (!channelInList && event.channel_type) {
             // If channel doesn't exist in existing list, check in activeChannels as well.
             // It may happen that channel was hidden using channel.hide(). In that case
             // We remove it from `channels` state, but its still being watched and exists in client.activeChannels.
             const channel = client.channel(event.channel_type, event.channel_id);
-
-            if (channel.initialized) {
-              return [channel, ...channels];
-            }
+            return uniqBy([channel, ...channels], 'cid');
```
Comment (santhoshvai) [inline package/src/components/ChannelList/hooks/listeners/useNewMessage.ts:45]: uniq is unnecessaruy isnt it? we are adding a new channel here..

File: package/src/components/ChannelList/hooks/listeners/useNewMessage.ts (hunk 2 above)
Comment (khushal87) [reply package/src/components/ChannelList/hooks/listeners/useNewMessage.ts:37]: No, it wasn't a bug. I have just aligned it the same as React SDK, and now it looks better and cleaner, IMO.

File: package/src/components/ChannelList/hooks/listeners/useNewMessage.ts (hunk 3 above)
Comment (khushal87) [reply package/src/components/ChannelList/hooks/listeners/useNewMessage.ts]: Fixed

File: package/src/components/Channel/Channel.tsx (hunk 1 above)
Comment (khushal87) [reply package/src/components/Channel/Channel.tsx]: Done

**Review by santhoshvai — APPROVED** [review]

File: package/src/components/ChannelList/utils.ts [hunk 4]
```diff
@@ -24,7 +25,7 @@ export const moveChannelUp = <
   channels.splice(index, 1);
   channels.unshift(channel);
 
-  return [...channels];
+  return uniqBy([channel, ...channels], 'cid');
```
[the remark this answers is not in the document]
Comment (khushal87) [reply package/src/components/ChannelList/utils.ts:28]: We need it else the channellist will have new channel as well as the old one 

simulator_screenshot_A14F91E1-8D58-4F21-BB0D-97549F55611C

**Pull request merged**